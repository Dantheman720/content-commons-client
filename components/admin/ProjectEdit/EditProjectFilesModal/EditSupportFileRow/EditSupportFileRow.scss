@import "styles/colors.scss";

.filename {
  [tooltip]::after {
    min-width: 220px;
  }
}
