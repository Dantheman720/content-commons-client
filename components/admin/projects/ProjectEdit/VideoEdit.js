import React from 'react';

const VideoEdit = () => (
  <div>Video Edit</div>
);

export default VideoEdit;
