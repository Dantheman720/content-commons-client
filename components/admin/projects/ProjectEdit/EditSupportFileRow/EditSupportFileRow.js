/**
 *
 * EditSupportFileRow
 *
 */
import React, { Fragment } from 'react';
import { array, func, object } from 'prop-types';
import {
  Button, Dropdown, Popup, Table
} from 'semantic-ui-react';
import gql from 'graphql-tag';
import { graphql, Query } from 'react-apollo';
import debounce from 'lodash/debounce';

import Focusable from 'components/admin/projects/shared/Focusable/Focusable';
import VisuallyHidden from 'components/admin/projects/shared/VisuallyHidden/VisuallyHidden';

import { SUPPORT_FILES_QUERY } from 'components/admin/projects/ProjectEdit/EditSupportFilesContent/EditSupportFilesContent';

import './EditSupportFileRow.scss';

const LANGUAGES_QUERY = gql`
  query Languages($orderBy: LanguageOrderByInput) {
    languages(orderBy: $orderBy) {
      ## use key alias to match
      ## Semantic UI property value
      key: id
      value: displayName
      text: displayName
    }
  }
`;

/* eslint-disable react/prefer-stateless-function */
class EditSupportFileRow extends React.PureComponent {
  constructor( props ) {
    super( props );

    this.DELAY_INTERVAL = 1000;
    this.STR_INDEX_PROPORTION = 0.04;
    this.ITEM_NAME_PROPORTION = 0.85;
    this.debounceResize = debounce( this.resetWidths, this.DELAY_INTERVAL );
    this._isMounted = false;

    this.state = {
      cellWidth: null,
      fileNameWidth: null,
      fileLanguage: ''
    };
  }

  componentDidMount = () => {
    this._isMounted = true;
    window.addEventListener( 'resize', this.debounceResize );
    this.setState( {
      fileLanguage: this.props.file.language.displayName
    } );
  }

  componentWillUnmount = () => {
    this._isMounted = false;
    window.removeEventListener( 'resize', this.debounceResize );
  }

  /**
   * Truncates long strings with ellipsis
   * @param {string} str the string
   * @param {number} start index for first cutoff point
   * @param {number} end index for ending cutoff point
   * @return truncated string
   */
  getShortFileName = ( str, index ) => (
    <Fragment>
      { `${str.substr( 0, index )}` }&hellip;{ `${str.substr( -index )}` }
    </Fragment>
  );

  /**
   * Determines an integer proportional
   * to a reference number
   * @param {number} reference
   * @return {number}
   */
  getProportionalNumber = ( reference, proportion ) => (
    Math.floor( reference * proportion )
  )

  setReplaceFileRef = input => {
    this.addReplaceFileRef = input;
  }

  /**
   * Declares a React ref & sets its width in state
   * @param {node} React node
   * @param {string} name of React ref
   */
  setRefWidth = ( node, ref ) => {
    if ( node ) {
      this.setState( prevState => {
        if ( !prevState[`${ref}Width`] ) {
          return ( {
            [`${ref}Width`]: Math.ceil( node.offsetWidth )
          } );
        }
      } );
    }
  }

  resetWidths = () => {
    if ( this._isMounted ) {
      this.setState( {
        cellWidth: null,
        fileNameWidth: null
      } );
    }
  }

  isLongName = ( itemWidth, reference, proportion ) => (
    itemWidth >= this.getProportionalNumber( reference, proportion )
  );

  handleChange = ( e, { value } ) => (
    this.setState( {
      fileLanguage: value
    } )
  )

  handleReplaceFile = () => {
    console.log( 'replace file' );
    this.addReplaceFileRef.click();
  }

  handleDeleteFile = () => {
    const { file: { id }, mutate } = this.props;
    mutate( { variables: { id } } );
  }

  renderIcons = () => {
    const { fileExtensions } = this.props;
    const isSingleType = fileExtensions.length === 1;
    const acceptedTypes = isSingleType ? fileExtensions[0] : '';

    return (
      <Button.Group basic size="large">
        <Popup
          content="Replace"
          size="mini"
          inverted
          on={ [
            'hover',
            'click',
            'focus'
          ] }
          trigger={ (
            <Button
              className="replace"
              icon="refresh"
              onClick={ this.handleReplaceFile }
              basic
              aria-label="replace file"
            />
          ) }
        />
        <VisuallyHidden>
          { /* eslint-disable jsx-a11y/label-has-for */
            /* eslint-disable jsx-a11y/label-has-associated-control */ }
          <label htmlFor="upload-file--single">upload file</label>
          <input
            id="upload-file--single"
            ref={ this.setReplaceFileRef }
            type="file"
            accept={ acceptedTypes }
            tabIndex={ -1 }
          />
        </VisuallyHidden>
        <Popup
          content="Delete"
          size="mini"
          inverted
          on={ [
            'hover',
            'click',
            'focus'
          ] }
          trigger={ (
            <Button
              className="delete"
              icon="delete"
              onClick={ this.handleDeleteFile }
              basic
              aria-label="delete file"
            />
          ) }
        />
      </Button.Group>
    );
  }

  render() {
    const { file, file: { filename, id } } = this.props;

    if ( !file || !Object.keys( file ).length ) return null;

    const { cellWidth, fileNameWidth, fileLanguage } = this.state;

    const charIndex = this.getProportionalNumber( fileNameWidth, this.STR_INDEX_PROPORTION );

    const shortFileName = this.getShortFileName( filename, charIndex );

    const isLongFileName = this.isLongName( fileNameWidth, cellWidth, this.ITEM_NAME_PROPORTION );

    const popupStyle = {
      overflowWrap: 'break-word',
      wordWrap: 'break-word',
      msWordBreak: 'break-all',
      wordBreak: 'break-word'
    };

    return (
      <Table.Row>
        <Table.Cell>
          <div
            className="file-name"
            ref={ node => this.setRefWidth( node, 'cell' ) }
          >
            { isLongFileName && <VisuallyHidden>{ filename }</VisuallyHidden> }
            <span
              className={
                `file-name-wrap${isLongFileName ? ' hasEllipsis' : ''}`
              }
              aria-hidden={ isLongFileName }
              ref={ node => this.setRefWidth( node, 'fileName' ) }
            >
              { isLongFileName
                ? (
                  <Popup
                    trigger={ (
                      <span>
                        <Focusable>{ shortFileName }</Focusable>
                      </span>
                    ) }
                    content={ filename }
                    on={ [
                      'hover', 'click', 'focus'
                    ] }
                    inverted
                    size="mini"
                    style={ popupStyle }
                  />
                )
                : filename }
            </span>
          </div>
        </Table.Cell>

        <Table.Cell>
          { /* eslint-disable jsx-a11y/label-has-for */
            <VisuallyHidden>
              <label htmlFor={ `file-${id}` }>
                { `${filename} language` }
              </label>
            </VisuallyHidden> }

          <Query query={ LANGUAGES_QUERY } variables={ { orderBy: 'displayName_ASC' } }>
            { ( { error, loading, data } ) => {
              if ( loading ) return 'Loading...';
              if ( error ) return `Error! ${error.message}`;

              return (
                <Dropdown
                  id={ id }
                  onChange={ this.handleChange }
                  options={ data.languages }
                  placeholder="â€“"
                  text={ fileLanguage }
                  value={ fileLanguage }
                  fluid
                  required
                  selection
                />
              );
            } }
          </Query>
        </Table.Cell>

        <Table.Cell>{ this.renderIcons() }</Table.Cell>
      </Table.Row>
    );
  }
}

EditSupportFileRow.propTypes = {
  file: object.isRequired,
  fileExtensions: array,
  mutate: func
};

const DELETE_SUPPORT_FILE_MUTATION = gql`
  mutation DeleteSupportFile($id: ID!) {
    deleteSupportFile(id: $id) {
      id
      filename
    }
  }
`;

export default graphql( DELETE_SUPPORT_FILE_MUTATION, {
  options: props => ( {
    refetchQueries: [
      {
        query: SUPPORT_FILES_QUERY,
        variables: { id: props.projectId }
      },
    ],
  } ),
} )( EditSupportFileRow );
export { DELETE_SUPPORT_FILE_MUTATION };
