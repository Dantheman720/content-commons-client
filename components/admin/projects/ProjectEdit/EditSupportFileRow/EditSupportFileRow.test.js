import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { projects } from 'components/admin/projects/ProjectEdit/mockData';
import EditSupportFileRow from './EditSupportFileRow';

const props = {
  handleChange: jest.fn(),
  file: projects[0].supportFiles.other[0],
  fileExtensions: ['.jpg', '.mp3'],
  selectedLanguage: projects[0].supportFiles.other[0].lang
};

const languages = [
  { value: 'arabic', text: 'Arabic' },
  { value: 'chinese', text: 'Chinese' },
  { value: 'english', text: 'English' },
  { value: 'french', text: 'French' },
  { value: 'portuguese', text: 'Portuguese' },
  { value: 'russian', text: 'Russian' },
  { value: 'spanish', text: 'Spanish' }
];

const resizeWindow = ( x, y ) => {
  window.innerWidth = x;
  window.innerHeight = y;
  window.dispatchEvent( new Event( 'resize' ) );
};

const Component = <EditSupportFileRow { ...props } />;

describe( '<EditSupportFileRow />', () => {
  it( 'renders without crashing', () => {
    const wrapper = shallow( Component );
    expect( wrapper.exists() ).toEqual( true );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( '`componentDidMount` sets `_isMounted`', () => {
    const wrapper = shallow( Component );
    wrapper.instance().componentDidMount();
    expect( wrapper.instance()._isMounted ).toEqual( true );
  } );

  it( '`componentWillUnmount` sets `_isMounted` to false', () => {
    const wrapper = shallow( Component );
    wrapper.instance().componentWillUnmount();
    expect( wrapper.instance()._isMounted ).toEqual( false );
  } );

  it( 'calls `debounceResize` when window is resized', () => {
    const wrapper = shallow( Component );
    const spy = jest.spyOn( wrapper.instance(), 'debounceResize' );

    wrapper.instance().componentDidMount();

    const map = {};
    window.addEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    resizeWindow( 900, 600 );
    expect( spy ).toHaveBeenCalled();
  } );

  it( '`componentWillUnmount` removes the `resize` event listener', () => {
    const wrapper = shallow( Component );
    const inst = wrapper.instance();
    const map = {};
    window.removeEventListener = jest.fn( ( event, cb ) => {
      map[event] = cb;
    } );

    wrapper.instance().componentWillUnmount();
    expect( window.removeEventListener )
      .toHaveBeenCalledWith( 'resize', inst.debounceResize );
  } );

  it( '`getShortFileName` returns an ellipsis shortened file name', () => {
    const wrapper = shallow( Component );
    const longName = props.file.file; // madeinamerica_arabic.jpg
    const index = 8;
    const shortName = wrapper.instance()
      .getShortFileName( longName, index )
      .props
      .children;

    expect( shortName ).toHaveLength( 3 );
    expect( shortName[0] ).toHaveLength( 8 );
    expect( shortName[1] ).toHaveLength( 1 );
    expect( shortName[2] ).toHaveLength( 8 );
    expect( shortName.join( '' ) ).toEqual( 'madeinamâ€¦abic.jpg' );
  } );

  it( '`getProportionalNumber` returns a number that is proportional to a reference', () => {
    const wrapper = shallow( Component );
    const reference = 500;
    const proportion = wrapper.instance().ITEM_NAME_PROPORTION;
    const num = wrapper.instance()
      .getProportionalNumber( reference, proportion );
    expect( num ).toEqual( 425 );
  } );

  it( '`isLongName` returns a boolean whether given number is greater or equal to another', () => {
    const wrapper = shallow( Component );
    let itemWidth = 290;
    const reference = 500;
    const proportion = wrapper.instance().ITEM_NAME_PROPORTION;

    const isLongName = () => (
      wrapper.instance()
        .isLongName( itemWidth, reference, proportion )
    );

    expect( isLongName() ).toEqual( false );
    itemWidth = 450;
    expect( isLongName() ).toEqual( true );
  } );

  it( '`setRefWidth` and `resetWidths` sets ref widths in state', () => {
    const wrapper = shallow( Component );
    const div = wrapper.find( '.file-name' );
    const span = wrapper.find( '.file-name-wrap' );

    const nodes = [div, span];
    const names = ['cell', 'fileName'];

    div.offsetWidth = 200;
    span.offsetWidth = 170;

    nodes.forEach( ( n, i ) => {
      expect( wrapper.state( `${names[i]}Width` ) )
        .toEqual( null );

      wrapper.instance().setRefWidth( n, names[i] );
      expect( wrapper.state( `${names[i]}Width` ) )
        .toEqual( n.offsetWidth );

      wrapper.instance().resetWidths();
      expect( wrapper.state( `${names[i]}Width` ) )
        .toEqual( null );
    } );
  } );

  it( '`resetWidths` does not reset ref widths in state if component is unmounted', () => {
    const wrapper = shallow( Component );
    const div = wrapper.find( '.file-name' );
    const span = wrapper.find( '.file-name-wrap' );

    const nodes = [div, span];
    const names = ['cell', 'fileName'];

    div.offsetWidth = 200;
    span.offsetWidth = 170;

    wrapper.instance().componentDidMount();
    const isMounted = () => wrapper.instance()._isMounted;

    expect( isMounted() ).toEqual( true );

    nodes.forEach( ( n, i ) => {
      expect( wrapper.state( `${names[i]}Width` ) )
        .toEqual( null );

      wrapper.instance().setRefWidth( n, names[i] );
      expect( wrapper.state( `${names[i]}Width` ) )
        .toEqual( n.offsetWidth );
    } );

    wrapper.instance().componentWillUnmount();
    expect( isMounted() ).toEqual( false );

    wrapper.instance().resetWidths();

    nodes.forEach( ( n, i ) => {
      expect( wrapper.state( `${names[i]}Width` ) )
        .toEqual( n.offsetWidth );
    } );
  } );

  it( 'renders `setReplaceFileRef`', () => {
    /**
     * Wrap Component in `<table><tbody></tbody></table>`
     * to prevent this terminal console warning:
     *
     * `Warning: validateDOMNesting(...): <tr> cannot appear as a child of <div>.`
     *
     * `mount` instead of `shallow` to get access to `ref`.
     */
    const wrapper = mount(
      <table>
        <tbody>
          <EditSupportFileRow { ...props } />
        </tbody>
      </table>
    );
    const row = wrapper.find( 'EditSupportFileRow' );
    const input = wrapper.find( 'input#upload-file--single' );
    expect( input.exists() ).toEqual( true );
    expect( row.instance().setReplaceFileRef ).toBeTruthy();
  } );

  it( 'sets correct `acceptedTypes` for the file upload `input`', () => {
    const wrapper = shallow( Component );
    const input = () => wrapper.find( 'input#upload-file--single' );

    expect( input().prop( 'accept' ) ).toEqual( '' );
    wrapper.setProps( { fileExtensions: ['.srt'], fileType: 'srt' } );
    expect( input().prop( 'accept' ) ).toEqual( '.srt' );
  } );

  it( '`null` or `{}` `file` does not crash the component', () => {
    const wrapper = shallow( Component );

    wrapper.setProps( { file: null } );
    expect( wrapper.html() ).toEqual( null );
    wrapper.setProps( { file: {} } );
    expect( wrapper.html() ).toEqual( null );
  } );

  it( 'selecting a language from the Dropdown calls `handleChange`', () => {
    const wrapper = shallow( Component );
    const dropdown = wrapper.find( 'Dropdown' );

    dropdown.simulate( 'change' );
    expect( props.handleChange ).toHaveBeenCalled();
  } );

  it( 'renders Dropdown with correct `options`', () => {
    const wrapper = shallow( Component );
    const dropdown = wrapper.find( 'Dropdown' );

    expect( dropdown.prop( 'options' ) ).toEqual( languages );
  } );

  it( 'renders Dropdown with correct `text` and `value`', () => {
    const wrapper = shallow( Component );
    const dropdown = wrapper.find( 'Dropdown' );
    const { selectedLanguage } = props;

    expect( dropdown.prop( 'text' ) ).toEqual( selectedLanguage );
    expect( dropdown.prop( 'value' ) ).toEqual( selectedLanguage );
  } );

  it( 'renders Dropdown with error if no `selectedLanguage`', () => {
    const newProps = { ...props, selectedLanguage: '' };
    const wrapper = shallow( <EditSupportFileRow { ...newProps } /> );
    const dropdown = wrapper.find( 'Dropdown' );

    expect( dropdown.prop( 'error' ) ).toEqual( !newProps.selectedLanguage );
  } );

  it( 'Popup is rendered if file name is long', () => {
    const wrapper = shallow( Component );
    const popup = () => (
      wrapper.find( `Popup[content="${props.file.file}"]` )
    );

    wrapper.setState( { cellWidth: 400, fileNameWidth: 350 } );
    expect( popup().exists() ).toEqual( true );
    wrapper.setState( { cellWidth: 400, fileNameWidth: 300 } );
    expect( popup().exists() ).toEqual( false );
  } );

  it( '`fileName` only (i.e., no Popup) is rendered if file name is not long', () => {
    const wrapper = shallow( Component );
    const span = () => wrapper.find( '.file-name-wrap' );
    const popup = () => (
      wrapper.find( `Popup[content="${props.file.file}"]` )
    );

    wrapper.setState( { cellWidth: 400, fileNameWidth: 300 } );
    expect( popup().exists() ).toEqual( false );
    expect( span().children().text() ).toEqual( props.file.file );
  } );

  // it( 'clicking Delete icon calls `handleDeleteFile`', () => {
  //   const wrapper = shallow( Component );
  //   const popup = wrapper.find( 'Popup[content="Delete"]' );
  //   const deleteBtn = shallow( popup.prop( 'trigger' ) );
  //   const spy = jest.spyOn( wrapper.instance(), 'handleDeleteFile' );

  //   deleteBtn.simulate( 'click' );
  //   expect( spy ).toHaveBeenCalled();
  // } );

  // it( 'clicking Replace icon calls `handleReplaceFile`', () => {
  //   const wrapper = mount(
  //     <table>
  //       <tbody>
  //         <EditSupportFileRow { ...props } />
  //       </tbody>
  //     </table>
  //   );
  //   const row = wrapper.find( 'EditSupportFileRow' );
  //   const popup = row.find( 'Popup[content="Replace"]' );
  //   const replaceBtn = shallow( popup.prop( 'trigger' ) );
  //   const spy = jest.spyOn( row.instance(), 'handleReplaceFile' );

  //   replaceBtn.simulate( 'click' );
  //   expect( spy ).toHaveBeenCalled();
  // } );
} );
