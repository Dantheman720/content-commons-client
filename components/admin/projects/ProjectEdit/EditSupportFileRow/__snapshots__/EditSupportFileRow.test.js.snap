// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditSupportFileRow /> renders without crashing 1`] = `
<TableRow
  as="tr"
  cellAs="td"
>
  <TableCell
    as="td"
  >
    <div
      className="file-name"
    >
      <VisuallyHidden>
        madeinamerica_arabic.jpg
      </VisuallyHidden>
      <span
        aria-hidden={true}
        className="file-name-wrap hasEllipsis"
      >
        <Popup
          content="madeinamerica_arabic.jpg"
          inverted={true}
          keepInViewPort={true}
          on={
            Array [
              "hover",
              "click",
              "focus",
            ]
          }
          position="top left"
          size="mini"
          style={
            Object {
              "msWordBreak": "break-all",
              "overflowWrap": "break-word",
              "wordBreak": "break-word",
              "wordWrap": "break-word",
            }
          }
          trigger={
            <span>
              <Focusable
                order={0}
              >
                <React.Fragment>
                  
                  …
                  madeinamerica_arabic.jpg
                </React.Fragment>
              </Focusable>
            </span>
          }
        />
      </span>
    </div>
  </TableCell>
  <TableCell
    as="td"
  >
    <VisuallyHidden>
      <label
        htmlFor="file-5682"
      >
        madeinamerica_arabic.jpg language
      </label>
    </VisuallyHidden>
    <Dropdown
      additionLabel="Add "
      additionPosition="top"
      closeOnBlur={true}
      deburr={false}
      error={false}
      fluid={true}
      icon="dropdown"
      id="5682"
      minCharacters={1}
      noResultsMessage="No results found."
      onChange={[MockFunction]}
      openOnFocus={true}
      options={
        Array [
          Object {
            "text": "Arabic",
            "value": "arabic",
          },
          Object {
            "text": "Chinese",
            "value": "chinese",
          },
          Object {
            "text": "English",
            "value": "english",
          },
          Object {
            "text": "French",
            "value": "french",
          },
          Object {
            "text": "Portuguese",
            "value": "portuguese",
          },
          Object {
            "text": "Russian",
            "value": "russian",
          },
          Object {
            "text": "Spanish",
            "value": "spanish",
          },
        ]
      }
      placeholder="–"
      renderLabel={[Function]}
      required={true}
      searchInput="text"
      selectOnBlur={true}
      selectOnNavigation={true}
      selection={true}
      text="Arabic"
      value="Arabic"
      wrapSelection={true}
    />
  </TableCell>
  <TableCell
    as="td"
  >
    <ButtonGroup
      basic={true}
      size="large"
    >
      <Popup
        content="Replace"
        inverted={true}
        keepInViewPort={true}
        on={
          Array [
            "hover",
            "click",
            "focus",
          ]
        }
        position="top left"
        size="mini"
        trigger={
          <Button
            aria-label="replace file"
            as="button"
            basic={true}
            className="replace"
            icon="refresh"
            onClick={[Function]}
          />
        }
      />
      <VisuallyHidden>
        <label
          htmlFor="upload-file--single"
        >
          upload file
        </label>
        <input
          accept=""
          id="upload-file--single"
          tabIndex={-1}
          type="file"
        />
      </VisuallyHidden>
      <Popup
        content="Delete"
        inverted={true}
        keepInViewPort={true}
        on={
          Array [
            "hover",
            "click",
            "focus",
          ]
        }
        position="top left"
        size="mini"
        trigger={
          <Button
            aria-label="delete file"
            as="button"
            basic={true}
            className="delete"
            icon="delete"
            onClick={[Function]}
          />
        }
      />
    </ButtonGroup>
  </TableCell>
</TableRow>
`;
